### DISCUSSION TITLE
Timeline Improvements

### DISCUSSION THREAD

**Discussion by OverloadedOrama** [Orama-Interactive/Pixelorama#587] [Ideas]
**Feature/enhancement description:**
v0.7 came with a new timeline, but I think there is still a lot of room for improvement. Some ideas that I have are:

- [x] Frame-specific delay.  (Issue #267 - implemented in #357)
- [x] Drag and drop individual cels to move them around. (27852b2694537287556e1e943428868e5572e4c4)
- [x] Hold <kbd>Control</kbd> and mouse wheel to resize cels in the timeline. (7f1594e1bc890dc29f8dad59f1aa2c0a23454974)
- [x] Create a button for each for frame instead of a label. Left clicking it selects the frame, right clicking it opens a menu with options suck as remove, clone, move & properties. (39b0baca84c0f4e27a3f593255edaf9085be2002 and ecd08ce132355c07d261b1867085dc09d0ea13fc)
- [x] Change the right click menu of the cels. It's currently frame-related, while it should be cel-related. (ecd08ce132355c07d261b1867085dc09d0ea13fc)
- [x] Clicking on a layer button selects the layer. Double clicking it allows renaming. (e7a5b4553611483a51fc8eb5d1902a5cd55040c2)
- [x] Put Delete and Link/Unlink in the cel right click menu. (1a87321437e9dc8da4b2415bbe71a39d02fad01c)
- [x] Drag and drop entire frames and layers to move them around. Their cels will also get moved with them. (e9bf310934d2bbc0cdb2b35dfba6af00e530f366)
- [x] Press <kbd>Control</kbd> to select multiple cels, and <kbd>Shift</kbd> to select multiple cels along with the cels between them. Pressing <kbd>Control</kbd> a on a selected cel would de-select it, unless it's the only selected cel. (de850ce8a11299f263fb867de7ecc3c792783fc3)
- [x] Changes made to a selected cel will also affect all other selected cels. That includes tools (like drawing with pencil) and image effects. (de850ce8a11299f263fb867de7ecc3c792783fc3)

EDIT: More improvements
- [x] Clone and delete multiple frames at once. (#648)
- [x] Adding a new frame inside a tag makes the new tag a part of the tag. (#648)
- [x] Adding a new frame outside of a tag to the left displaces the tag to the right. (#648)
- [x] Figure out where to put cel opacity and layer opacity. Layer opacity does not exist yet (359f509d57e710acff897b4535dc080d3c20031e).

The list may be edited to include more ideas in the future.

**Comment by novhack** [comment]
I have a big monitor and I would really like an option to resize cels. :slightly_smiling_face:

**Comment by Erevoid** [comment]
Got some more ideas regarding **tags**:

- [x] The ability to move the tag with its cels left or right in the timeline (in regards to other tags or single untagged frames)
- [ ] By right clicking the tag name there can be a popup with tag properties (to change the delay to each cel in all of the tags) and
- [x] The ability to clone all of the cels of a tag and make a cloned tag with [tagname]_2 on its name by default.

Those changes make the tags much more valuable and can save a lot of animation time.

**Comment by OverloadedOrama** [comment]
https://user-images.githubusercontent.com/35376950/106492160-7b21f280-646c-11eb-8808-d321d2ea68cb.mp4

Cel drag and dropping and cel resizing have been implemented and will included in the next version.

**Comment by Erevoid** [comment]
Some more improvement ideas:
- [x] You should be able to copy and paste multiple frames, or even copy the frames of a tag (clone the tag as mentioned in my comment above).
- [x] If you have 30 frames let's say, and multiple tags, covering all the frames, when you create a new frame from frame 1, every other frame is moved to the right, meaning all of the tags are not relevant anymore. The newest frame should either be part of the tag it was created in, or it should be displaced just to the right of that tag (outside of all the tags) and the rest of the tags on the right of the new frame should be displaced as well by 1 (if the tag used to cover 10-13 now it should cover 11-14). 
- [x] Similarly, if you have multiple frames and multiple tags covering them all and you clone a frame, the tag of the cloned frame is expanded and intersects with the tag on the right which creates confusion. The tags on the right should be displaced.
- [ ] While viewing the animation of some frames under a specific tag (after having checked the animation plays only on frames of the same tag), if you press a frame of another tag you view it for a fraction of a second and then it goes back to showing you the frames of the initial tag. It would be better if it continued playing the animation of the second tag, starting from the frame you just clicked, instead of the first one, since the user clicked the other frame for a reason.
- [ ] The tag UI is bad. By default only two tags are visible at all times unless you resize it. The naming "tag 1, tag 2" etc is not needed, since the user named the tag already, it should be the name instead. The edit tag button should be at the same line with the tag name (which will be if the name replaces tag1, tag2). Instead of Tag 1 it could look like this: 1 [Tag name] [frames] [edit button]. The add new tag button should always be visible without the need to scroll to find it, like the "animation plays only on frames of the same tag" is.
- [ ] Linked/Unlinked behaviour is weird to use. If I have many frames and then decide to link them I have to do it one by one

**Comment by novhack** [comment]
I think that there should be shortcut buttons to move between frames and layers.
